src/Analyser/AnalysedFilesResolver.php

```diff
@@ -0,0 +1,47 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Analyser;
+
+use function array_fill_keys;
+use function array_map;
+use function strtolower;
+use const PHP_OS_FAMILY;
+
+final class AnalysedFilesResolver
+{
+
+	/** @var bool[] filePath(string) => bool(true) */
+	private array $analysedFiles;
+
+	private bool $caseInsensitiveFilesystem;
+
+	/**
+	 * @param string[] $files
+	 */
+	public function __construct(array $files = [])
+	{
+		$this->caseInsensitiveFilesystem = PHP_OS_FAMILY === 'Darwin';
+		$this->setAnalysedFiles($files);
+	}
+
+	/**
+	 * @param string[] $files
+	 */
+	public function setAnalysedFiles(array $files): void
+	{
+		if ($this->caseInsensitiveFilesystem) {
+			$files = array_map(static fn (string $file): string => strtolower($file), $files);
+		}
+		$this->analysedFiles = array_fill_keys($files, true);
+	}
+
+	public function isInAnalyzedFiles(string $file): bool
+	{
+		if ($this->caseInsensitiveFilesystem) {
+			$file = strtolower($file);
+		}
+
+		return isset($this->analysedFiles[$file]);
+	}
```

Comment:
actual fix is here, everything else are adjustments to make sure this new logic is used all over the place